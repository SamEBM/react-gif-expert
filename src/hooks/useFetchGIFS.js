import { useEffect, useState } from "react";
import { getGIFS } from '../helpers/getGIFS';

export const useFetchGIFS = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
        getGIFS(category).then(gifsResult => {

            setTimeout(() => {
                setState({
                    data: gifsResult,
                    loading: false
                });
            }, 2000);
        });
    }, [category]);

    return state;
}