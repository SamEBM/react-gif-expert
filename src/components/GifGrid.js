// import React, { useEffect, useState } from 'react'
// import { getGIFS } from '../helpers/getGIFS';
import GifItem from './GifItem';
import {useFetchGIFS} from '../hooks/useFetchGIFS';

const GifGrid = ({ category }) => {
    // const [gifs, setGifs] = useState([]);

    // Custom Hook
    const { data : images, loading } = useFetchGIFS(category);

    // Esta funcion se ejecuta la primera vez que se renderiza el componente y si 'category' llega a cambiar
    // useEffect(() => {
    //     getGIFS(category).then(gifsResult => {
    //         setGifs(gifsResult);
    //     });
    // }, [category]);

    return (
        <>
            <h3 className='animate__animated animate__backInLeft'>{category}</h3>
            { loading && <p className='animate__animated animate__pulse'>Cargando...</p>}
            <div className='card-grid'>
                {
                    images.map( gif => {
                        return <GifItem key={gif.id} {...gif}/>
                    })
                }
            </div>
        </>
    );
}

GifGrid.propTypes = {}

export default GifGrid