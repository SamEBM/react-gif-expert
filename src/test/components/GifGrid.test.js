import { shallow } from 'enzyme';
import GifGrid from '../../components/GifGrid';
import { useFetchGIFS } from '../../hooks/useFetchGIFS';
jest.mock('../../hooks/useFetchGIFS');

describe('Pruebas con Gif Grid', () => { 
    const category = 'Attack on Titan';

    // Esto se ejecuta cada vez que se ejecuta un test
    // beforeEach( () => {
    //     jest.clearAllMocks();
    //     wrapper = shallow(<GifGrid category={category}/>);
    // });

    test('Debe mostrar <GifGrid /> correctamente', () => { 

        //Simular como si fuese la inicializacion del componente 
        useFetchGIFS.mockReturnValue({
            data: [],
            loading: true
        });
        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe mostrar gifs cuando termina useFetchGifs', () => { 
        const gifs = [
            {
                id: 'ABC',
                url: 'https://localhost/culaquier1.gif',
                title: 'Cualquier cosa1'
            },
            {
                id: 'DEF',
                url: 'https://localhost/culaquier2.gif',
                title: 'Cualquier cosa2'
            },
        ];

        // Simular que ya se obtuvo respuesta del custom Hook 
        useFetchGIFS.mockReturnValue({
            data: gifs,
            loading: false
        });
        const wrapper = shallow(<GifGrid category={category}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifItem').length).toBe(gifs.length)
    })
})