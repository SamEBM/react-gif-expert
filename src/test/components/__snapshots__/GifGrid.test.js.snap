// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con Gif Grid Debe mostrar <GifGrid /> correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__backInLeft"
  >
    Attack on Titan
  </h3>
  <p
    className="animate__animated animate__pulse"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas con Gif Grid Debe mostrar gifs cuando termina useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__backInLeft"
  >
    Attack on Titan
  </h3>
  <div
    className="card-grid"
  >
    <GifItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa1"
      url="https://localhost/culaquier1.gif"
    />
    <GifItem
      id="DEF"
      key="DEF"
      title="Cualquier cosa2"
      url="https://localhost/culaquier2.gif"
    />
  </div>
</Fragment>
`;
